app/services/weather.py

```diff
@@ -0,0 +1,329 @@
+            },
+        }
+
+    async def get_weather_forecast(self, city: str, start_date: Any, end_date: Any) -> Dict[str, Any]:
+        """Fetch and normalize weather forecast for the given city and date range.
+
+        Args:
+            city: City name (e.g., "Kyiv", "Lviv").
+            start_date: Inclusive start date (str, datetime, or epoch).
+            end_date: Inclusive end date (str, datetime, or epoch).
+
+        Returns:
+            Normalized JSON dict with daily forecasts in Celsius. On failure, returns
+            an empty payload with a `disabled` or `error` note so the caller can
+            gracefully fall back.
+        """
+        
+        if not self._api_key:
+            return {"disabled": True, "reason": "missing_api_key"}
+
+        try:
+            start_dt = _coerce_date(start_date)
+            end_dt = _coerce_date(end_date)
+        except Exception as e:
+            return {"error": "invalid_dates"}
+
+        cache_hit = self._cache.get(city, start_dt.strftime("%Y-%m-%d"), end_dt.strftime("%Y-%m-%d"))
+        if cache_hit is not None:
+            return cache_hit
+
+        attempt = 0
+        backoff = self._retry_min
+        async with aiohttp.ClientSession() as session:
+            while True:
+                try:
+                    coords = await self._get_city_coordinates(session, city)
+                    if not coords:
+                        return {"error": "geocoding_failed"}
+                    lat, lon = coords
+                    raw = await self._fetch_forecast_raw(session, lat, lon)
+                    daily = _aggregate_to_daily(raw.get("list", []))
+                    sliced = self._slice_by_dates(daily, start_dt, end_dt)
+                    normalized = self._to_normalized(city, sliced)
+                    self._cache.set(city, start_dt.strftime("%Y-%m-%d"), end_dt.strftime("%Y-%m-%d"), normalized)
+                    return normalized
+                except Exception as e:
+                    attempt += 1
+                    if attempt > self._retry_attempts:
+                        logger.warning(f"Weather request failed after retries: {e}")
+                        return {"error": "request_failed"}
+                    await asyncio.sleep(min(self._retry_max, max(self._retry_min, backoff)))
+                    backoff = min(self._retry_max, backoff * 2)
+
+
+def get_weather_forecast_sync(project_root: str, city: str, start_date: Any, end_date: Any) -> Dict[str, Any]:
+    """Synchronous convenience wrapper for environments without asyncio plumbing."""
+    service = WeatherService(project_root=project_root)
+    try:
+        return asyncio.run(service.get_weather_forecast(city, start_date, end_date))
+    except RuntimeError:
+        # If already in an event loop (e.g., Streamlit), create a task in that loop
+        loop = asyncio.get_event_loop()
+        return loop.run_until_complete(service.get_weather_forecast(city, start_date, end_date))
```

Comment:
asyncio.run will fail if already inside an event loop, which can occur in Streamlit Use asyncio.run only if outside an event loop, else call create_task.